cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesTable.java

```diff
@@ -0,0 +1,196 @@
+    // if the value is null and the entity has a parent, defer to the parent
+    if (value == null && entityId instanceof ParentedId) {
+      value = getResolvedPreference(((ParentedId) entityId).getParent(), name);
+    }
+    // return whatever we have
+    return value;
+  }
+
+  /**
+   * Set the preferences for the entity id
+   *
+   * @param entityId the entity id to set the preferences from
+   * @param propMap the map which contains the preferences
+   */
+  public void setPreferences(EntityId entityId, Map<String, String> propMap) throws IOException {
+    switch (entityId.getEntityType()) {
+      case INSTANCE:
+        table.createOrUpdate(EMPTY_NAMESPACE, INSTANCE_PREFERENCE, new Config(entityId.getEntityName(), propMap));
+        break;
+      case NAMESPACE:
+        NamespaceId namespaceId = (NamespaceId) entityId;
+        table.createOrUpdate(
+          namespaceId.getNamespace(), NAMESPACE_PREFERENCE, new Config(namespaceId.getNamespace(), propMap));
+        break;
+      case APPLICATION:
+        ApplicationId appId = (ApplicationId) entityId;
+        table.createOrUpdate(
+          appId.getNamespace(), APPLICATION_PREFERENCE, new Config(appId.getApplication(), propMap));
+        break;
+      case PROGRAM:
+        ProgramId programId = (ProgramId) entityId;
+        table.createOrUpdate(
+          programId.getNamespace(), PROGRAM_PREFERENCE, new Config(getProgramName(programId), propMap));
+        break;
+      default:
+        throw new UnsupportedOperationException(
+          String.format("Preferences cannot be used on this entity type: %s", entityId.getEntityType()));
+    }
+  }
+
+  /**
+   * Delete the preferences for the entity id
+   *
+   * @param entityId the entity id to delete the preferences
+   */
+  public void deleteProperties(EntityId entityId) throws IOException {
+    try {
+      switch (entityId.getEntityType()) {
+        case INSTANCE:
+          table.delete(EMPTY_NAMESPACE, INSTANCE_PREFERENCE, entityId.getEntityName());
+          break;
+        case NAMESPACE:
+          NamespaceId namespaceId = (NamespaceId) entityId;
+          table.delete(namespaceId.getNamespace(), NAMESPACE_PREFERENCE, namespaceId.getNamespace());
+          break;
+        case APPLICATION:
+          ApplicationId appId = (ApplicationId) entityId;
+          table.delete(appId.getNamespace(), APPLICATION_PREFERENCE, appId.getApplication());
+          break;
+        case PROGRAM:
+          ProgramId programId = (ProgramId) entityId;
+          table.delete(programId.getNamespace(), PROGRAM_PREFERENCE, getProgramName(programId));
+          break;
+        default:
+          throw new UnsupportedOperationException(
+            String.format("Preferences cannot be used on this entity type: %s", entityId.getEntityType()));
+      }
+    } catch (ConfigNotFoundException e) {
+      //no-op
+    }
+  }
+
+  private String getProgramName(ProgramId programId) {
```

Comment:
Is this method here because program can same name but in different app or type? If so, it is better to add a comment why program is treated different than app and namespace.

Reply:
Done